File: OBI 2017/arranhaceu.py
def LSOne(S):
    return (S & (-S))

class FenwickTree:
    def __init__(self, n):
        self.__ft = [0] * (n + 1)

    def adjust(self, k, v):
        while k < len(self.__ft):
            self.__ft[k] += v
            k += LSOne(k)
    
    def rsq(self, b):
        soma = 0
        while b > 0:
            soma += self.__ft[b]
            b -= LSOne(b)
        return soma

N, Q = map(int, input().split())
v = list(map(int, input().split()))
bit = FenwickTree(N)
for k in range(N):
    bit.adjust(k + 1, v[k])
for _ in range(Q):
    dados = list(map(int, input().split()))
    q = dados[0]
    k = dados[1]
    if q == 1:
        print(bit.rsq(k))
    else:
        p = dados[2]
        bit.adjust(k, p - v[k-1])
        v[k-1] = p
        

File: OBI 2017/arranhaceu.cpp
#include <iostream>
#include <vector>
#include <string.h>
#define MAX 112345
using namespace std;
typedef long long ll;
int ft[MAX];

int LSOne(int S)
{
    return S & (-S);
}

void adjust(int k, ll v)
{
    while (k < MAX)
    {
        ft[k] += v;
        k += LSOne(k);
    }
}

ll rsq(int b)
{
    int soma = 0;
    while (b > 0)
    {
        soma += ft[b];
        b -= LSOne(b);
    }
    return soma;
}

int main()
{
    memset(ft, 0, MAX);
    int N, Q;
    cin >> N >> Q;
    vector<int> v;
    for (int i = 0; i < N; i++)
    {
        ll s;
        cin >> s;
        v.push_back(s);
        adjust(i + 1, s);
    }
    for (int i = 0; i < Q; i++)
    {
        int q, k;
        cin >> q >> k;
        if (q == 1)
        {
            cout << rsq(k) << endl;
        }
        else
        {
            ll val;
            cin >> val;
            adjust(k, val - v[k - 1]);
            v[k - 1] = val;
        }
    }

    return 0;
}